modules/3d/src/spc.cpp

```diff
@@ -0,0 +1,244 @@
+    sqrt(degree_mat, degree_mat);
+    Mat degree_mat_sqrt;
+    repeat(degree_mat, 1, in.rows, degree_mat_sqrt);
+    out = ((in.mul(degree_mat_sqrt).t()).mul(degree_mat_sqrt)).t();
+}
+
+void SpectralCluster::Impl::getAffinityMat(Mat &in, Mat &out, const std::vector<std::vector<int32_t>> &indices) {
+    int num_faces = int(indices.size());
+
+    // using dijkstra algorithm to get the affinity matrix
+    out = Mat::ones(num_faces, num_faces, CV_32F) * (double)std::numeric_limits<float>::infinity();
+    struct dist_node {
+        int index;
+        float distance;
+        dist_node(int i, float d) : index(i), distance(d) {}
+        bool operator<(const dist_node& other) const { return distance > other.distance; }
+    };
+
+    cv::parallel_for_(cv::Range(0, num_faces), [&](const cv::Range& range) {
+        for (int i = range.start; i < range.end; ++i) {
+            std::vector<bool> visited(num_faces, false);
+            std::priority_queue<dist_node> pq;
+            pq.emplace(i, 0.f);
+
+            while (!pq.empty()) {
+                dist_node current = pq.top();
+                pq.pop();
+
+                if (visited[current.index])
+                    continue;
+
+                visited[current.index] = true;
+
+                for (int neighbor = 0; neighbor < num_faces; ++neighbor) {
+                    float weight = in.at<float>(current.index, neighbor);
+                    if (weight > 0 && !visited[neighbor]) {
+                        float distance = current.distance + weight;
+                        if (distance < out.at<float>(i, neighbor)) {
+                            out.at<float>(i, neighbor) = distance;
+                            pq.emplace(neighbor, distance);
+                        }
+                    }
+                }
+            }
+        }
+    });
+
+    out.setTo(0, out == std::numeric_limits<float>::infinity());
+
+    Scalar sum = cv::sum(out);
+    auto sum_val = static_cast<float>(sum[0]);
+    float sigma = sum_val / static_cast<float>(num_faces * num_faces);
+    sigma = 2.0f * sigma * sigma;
+
+    // Gaussian Kernel
+    out.forEach<float>([sigma](float& elem, const int* position) -> void {
+        CV_UNUSED(position);
+        elem = (elem == 0.f) ? 0.f : exp(-elem / sigma);
+    });
+
+    // set diagonal elements to 1
+    for (int i = 0; i < num_faces; ++i)
+        out.at<float>(i, i) = 1.f;
+}
+
+void SpectralCluster::Impl::getAdjacentDistanceMat(Mat &out, const std::vector<Point3f> &vertices,
+                                             const std::vector<std::vector<int32_t>> &indices) const {
+    int num_faces = int(indices.size());
+    int num_adjacency = 0;
+    struct PairHash {
+        std::size_t operator() (const std::pair<int, int>& pair) const {
+            return std::hash<int>{}(pair.first) ^ std::hash<int>{}(pair.second);
+        }
+    };
+    std::unordered_map<std::pair<int, int>, std::pair<int, int>, PairHash> map;
```

Comment:
I propose different variable name; `map` can be confused with the STL container.